src/lib/packager/clients/pkg_finish.rb

```diff
@@ -0,0 +1,171 @@
+      # (needs to be done _after_ saving repositories, see bnc#700881)
+      Pkg.SourceCacheCopyTo(Installation.destdir)
+
+      # copy list of failed packages to installed system
+      WFM.Execute(
+        path(".local.bash"),
+        format("test -f %<path>s && /bin/cp -a %<path>s '%<destdir>s%<path>s'",
+          path: FAILED_PACKAGES_PATH, destdir: String.Quote(Installation.destdir)))
+    end
+
+  private
+
+    # During upgrade, old sources are reinitialized
+    # either as enabled or disabled.
+    # The old sources from targed should go away.
+    def BackUpAllTargetSources
+      Yast.import "Directory"
+
+      if !File.exist?(REPOS_DIR)
+        log.error("Directory #{REPOS_DIR} doesn't exist!")
+        return
+      end
+
+      current_repos = SCR.Read(path(".target.dir"), REPOS_DIR)
+
+      if current_repos.nil? || current_repos.empty?
+        log.warn("There are currently no repos in #{REPOS_DIR} conf dir")
+        return
+      else
+        log.info("These repos currently exist on a target: #{current_repos}")
+      end
+
+      cmd = WFM.Execute(path(".local.bash_output"), "date +%Y%m%d-%H%M%S")
+      a_name_list = (cmd["stdout"] || "the_latest").split("\n")
+      archive_name = "repos_#{a_name_list.first}.tgz"
+
+      compress_cmd = format(TAR_CMD,
+        source: String.Quote(Ops.add(Directory.vardir, "/repos.d_backup/")),
+        archive: String.Quote(archive_name),
+        target: String.Quote(REPOS_DIR))
+#
+      cmd = SCR.Execute(path(".target.bash_output"), compress_cmd)
+
+      if !cmd["exit"].zero?
+        log.error("Unable to backup current repos; Command >#{compress_cmd}< returned: #{cmd}")
+      end
+
+      current_repos.each do |repo|
+        file = File.join(REPOS_DIR, repo)
+        log.info("Removing target repository #{file}")
+        if !SCR.Execute(path(".target.remove"), file)
+          log.error("Cannot remove #{one_repo} file")
+        end
+      end
+
+      log.info("All old repositories were removed from the target")
+
+      # reload the target to sync the removed repositories with libzypp repomanager
+      Pkg.TargetFinish
+      Pkg.TargetInitialize("/mnt")
+
+      nil
+    end
+
+    # Disable given repositories if needed
+    #
+    # Given a repository with a scheme contained in 'schemes':
+    #
+    # * if all products it contains are available through another repository,
+    #   then it will be disabled;
+    # * if some product contained is not available through another
+    #   repository, then it will be left untouched.
+    #
+    # @param schemes [Array<Symbol>] Schemes to consider
+    # @return [Array<Packages::Repository>] List of disabled repositories
+    def disable_uneeded_repos(schemes)
+      candidates_repos, other_repos = *::Packages::Repository.enabled.partition do |repo|
+        schemes.include?(repo.scheme)
+      end
+      product_names = other_repos.map(&:products).flatten.map(&:name)
```

Comment:
So we just evaluate the plain product names? I still think we should stay on the safe side and at least check the version and possibly vendor.

Think about some strange corner cases, e.g. the SLES for SAP installation medium includes the original SLES product and additionally the SLES-for-SAP product, i.e. there are actually two (!) products there...